### PR TITLE
Added options to load dashboard on-the-fly

### PR DESCRIPTION
**Pull request by Simon-Free**
A PoC for the issue #112

### CODE REVIEW

**Comment by h4c5** [comment]
Example of use : 

```python
import os
import re

from explainerdashboard import ExplainerHub, ExplainerDashboard
from flask import request, redirect

current_path = os.path.dirname(os.path.abspath(__file__))

def load_dashboard_example(folder):
    """Load a dashboard from folder"""
     
    return ExplainerDashboard.from_config(
        os.path.join(current_path, folder, "explainer.joblib"), #f"{path}/explainer.joblib", 
        os.path.join(current_path, folder, "dashboard.yaml"), #f"{path}/config.yaml"
    )


if __name__ == "__main__":
    db = load_dashboard_example("model1")
    hub = ExplainerHub([db], max_dashboard=5)

    # We can now add a dashboard after the bus has been created
    db2 = load_dashboard_example("model1")
    hub.add_dashboard(db2)

    # When the number of dashboard reach max_dashboard the very first dashboards
    # are removed
    for i in range(10):
        db2 = load_dashboard_example("model1")
        hub.add_dashboard(db2)

    # Example of dashboard creation on the fly :
    # if a new dashboard is requested and we can find the corresponding folder
    # before answering request
    dashboard_pattern = re.compile(r"/dashboard/(\w+)/?")

    @hub.app.before_request
    def new_dashboard():
        """Create a new dashboard on the fly if possible"""
        
        match = dashboard_pattern.match(request.path)
        if match:
            dashboard_name, = match.groups()
            folder = os.path.join(current_path, dashboard_name)
            
            if dashboard_name not in hub.dashboard_names and os.path.exists(folder):
                db = load_dashboard_example(dashboard_name)
                
                db.name = dashboard_name
                hub.add_dashboard(db, bootstrap="/static/bootstrap.min.css")

                # Redirect to the dashboard page
                return redirect(f"/dashboard/_{dashboard_name}", code=302)

    hub.run(port=8050)
```

**Comment by oegedijk** [comment]
Hi, thanks so much! Looks cool!

Unfortunately, I will not have time until later next week to properly have a look at it.  But just from a glance I think I need to understand a bit better what problem you are trying to solve precisely.

I thought you wanted to solve the problem of lazy loading of dashboards, where you don't need to have all dashboards in memory all the time, but only load them from disk when a request comes in for that specific dashboard. 

But from a glance at the code, it seems you are adding the functionality to add new dashboards to an existing hub instance. And then I guess you use this to have dynamic routes that correspond to folder names and when there is a match and there is a dashboard.yaml present in that folder then dynamically add that dashboard to the hub?

**Comment by oegedijk** [comment]
So... why wouldn't you want to have all the dashboards on the index page from the start for easy discoverability? 

If I understand correctly here you would have to know the name of the folder in order to discover the dashboard? (which may be okay depending on your context, but in general part of the point of the ExplainerHub is to make the dashboards discoverable.)

**Comment by h4c5** [comment]
It is true it is a bit confusing without context. Our problem is that we want to add extra dashboard to a running hub without restarting the app.

The idea is to start the hub with all the dashboards stored in a directory. Then, while we train more models and create new ExplainerDashboard, add them in the directory and retrieve them in the running hub.

**Comment by oegedijk** [comment]
Okay, in that case maybe we should be a bit more ambitious and actually build an admin page where people with an admin role can add, delete or reload dashboards?

Would be a bit of work, and I don't have a lot of experience in this (data scientist not a web developer), but I think that would make sense.

In the short run however we could first build this shortcut where going to a route that corresponds to a subfolder automatically adds it to the hub.

**Comment by h4c5** [comment]
An admin could be great but it would be a bit of work indeed 😅

However it solves a different problem I think. Auto-creating dashboard from a repository is an easy way for us to combine ExplainerDashboard and MLFlow. 
We just add dashboard URL as MLFlow artifacts and when the URL is called, a dashboard is automatically integrated to our hub and served so it is accessible in MLFlow.

**Comment by oegedijk** [comment]
Ah, okay gotcha! Then I indeed can see the benefit of this approach. Will try to do a code review later this week and then get it merged and released. Could you then also write some documentation on how to combine it with MLFlow? I think that could be really useful to a lot of other people!

**Comment by oegedijk** [comment]
Okay, had a chance to go through the code, really cool work guys!

Made a few minor changes (e.g. max_dashboard->max_dashboards), and added an option `add_dashboard_route=True` to ExplainerHub that adds a Flask route `/add_dashboard/dashboard_name` to add new dashboards.

Example of use:

```python
db1 = ExplainerDashboard(explainer, title="First Dashboard")
db1.to_yaml("dashboards/dashboard1/dashboard.yaml", dump_explainer=True)

db2 = ExplainerDashboard(explainer, title="Second dashboard")
db2.to_yaml("dashboards/dashboard2/dashboard.yaml", dump_explainer=True)

hub = ExplainerHub([db1], max_dashboards=5, add_dashboard_route=True)
hub.run()
```

If you then go to `localhost:8050/add_dashboard/dashboard1` it should automatically add the second dashboard to the hub.

For now I am assuming that you store your dashboard in a subdirectory `dashboards/dashboard_name` and that` the file is called `dashboard.yaml`. Not sure whether should also make this configurable or if it is okay to be a bit opinionated on this.

Anything else we should add?

File: explainerdashboard/dashboards.py [hunk 1]
```diff
@@ -1030,11 +1036,8 @@ def __init__(self, dashboards:List[ExplainerDashboard], title:str="ExplainerHub"
             self.app.config['SECRET_KEY'] = secret_key
         SimpleLogin(self.app, login_checker=self._validate_user)
         
-        self.base_route = base_route
```
Comment (h4c5) [inline explainerdashboard/dashboards.py]: Should we keep this line ?

Comment (oegedijk) [reply explainerdashboard/dashboards.py]: No need. `self._store_params(no_store=['dashboards', 'logins', 'secret_key'])` automatically stores all parameters to attributes. (saves you a bunch of boilerplate)

**Comment by h4c5** [comment]
Cool, thank you for reviewing and for the changes !

We sure can make a doc for integration with MLFlow 😄 
Where should we place it ? 

About being opiniated on the dashboards directory, I rather prefer let the choice to the user.

File: explainerdashboard/dashboards.py (hunk 1 above)
Comment (h4c5) [reply explainerdashboard/dashboards.py]: Nice !

**Comment by oegedijk** [comment]
Here is the hub documentation if you want to have a stab: https://github.com/oegedijk/explainerdashboard/blob/master/docs/source/hub.rst

If you can add a section there then I can link to it in the README.

**Comment by oegedijk** [comment]
So I rewrote the`/add_dashboard` so that you specify an exact relative `.yaml` file, e.g:

```
\add_dashboard\dashboard3.yaml
\add_dashboard\dashboards\dashboard4\dashboard.yaml
```

Also added a `\remove_dashboard` route:

```
\remove_dashboard\db1
```

(this replaces the `\dashboards\db2\remove` route)

It seems that you can remove the dashboard from the index, but the old routes will remain registered. So if you go directly to `/dashboard/db2` you can still see the dashboard. (and it would still be taking up memory as well I guess). Not sure how to remove those. But also added some checks that exist if any old routes already exist, and if so add a counter to `dashboard.name`.

**Comment by Simon-Free** [comment]
We were wondering with @h4c5, if a possible solution to the hardcoded dashboard path could be to add an option to provide a mapping function (taking an URL, returning a filepath), with the default one having the behavior you were describing before 🤔

**Comment by oegedijk** [comment]
Yeah, I figured this way is even more flexible: simply allow to user to specify the filepath in the URL...

Do you see downsides from this approach? 

Alternative would be to e.g. add parameter `dashboard_folder='dashboards'` to ExplainerHub and then search in that folder. But I think the current design is nicely flexible...

**Comment by oegedijk** [comment]
One concern I have is that this will not scale when you deploy it across a cluster: only one of the pods would get updated.

Not sure there is any way around that. So maybe should just give a warning to the user.

**Comment by oegedijk** [comment]
Added some warnings about multiple workers/nodes. Think this is good to go for now? Will probably do a release tomorrow with some other stuff (limited PyTorch support, a simplified default dashboard), so could include this as well...

**Comment by h4c5** [comment]
> 
> 
> Yeah, I figured this way is even more flexible: simply allow to user to specify the filepath in the URL...
> 
> Do you see downsides from this approach?
> 
> Alternative would be to e.g. add parameter `dashboard_folder='dashboards'` to ExplainerHub and then search in that folder. But I think the current design is nicely flexible...

It could be nice indeed. Have you thought about the url pattern which could allow to do that ? 

Adding warnings is a good idea. In fact I would not know how to do anything better anyway. I do not know much about running a dash app in a cluster.

Yes, I think the feature is good enough for now. I would enjoy to see it in the next release 😄 

PS: Sorry for not responding any faster.

**Comment by oegedijk** [comment]
Added add_dashboard_pattern. Use:

```python
db1 = ExplainerDashboard(explainer, title="First Dashboard", name="db1")
db1.to_yaml("dashboards/dashboard1/dashboard.yaml", dump_explainer=True)

db2 = ExplainerDashboard(explainer, title="Second dashboard", name="db2")
db2.to_yaml("dashboards/dashboard2/dashboard.yaml", dump_explainer=True)

db3 = ExplainerDashboard(explainer, title="Third dashboard", name='doobydobodo')
db3.to_yaml("dashboard3.yaml", dump_explainer=True)

hub = ExplainerHub([db1], max_dashboards=5, 
                   add_dashboard_route=True, 
                   add_dashboard_pattern="dashboards/{}/dashboard.yaml")
```

So now `\add_dashboard\dashboard2` will work, but `\add_dashboard\dashboard3.yaml` no longer will.

**Comment by oegedijk** [comment]
Released it here: https://github.com/oegedijk/explainerdashboard/releases/tag/v0.3.5

Thanks again for the contribution!

**Comment by h4c5** [comment]
Thank you for your work !

**Comment by marcelo-mms** [comment]
Hi!

First, thanks for this cool library. It is really good.

I'm trying to follow these steps to add more dashboards on the fly but it is not working. I'm using Python 3.12 (also tried 3.11 in a docker container) and `explainerdashboard==0.4.7`. This is a dummy example I wrote:

```python
from explainerdashboard import ExplainerDashboard, RegressionExplainer, ExplainerHub, ClassifierExplainer
import numpy as np
import pandas as pd
from sklearn.tree import DecisionTreeRegressor, DecisionTreeClassifier


n = 100

df = pd.DataFrame(
    data=np.random.randint(low=0, high=5, size=(n, 10)),
    columns=list('abcdefghij')
).assign(
    target = np.random.random(n)
)

features = df.drop(columns=['target'])
target = df.target


model = DecisionTreeRegressor().fit(features, target)
explainer = RegressionExplainer(model, features, target)


db1 = ExplainerDashboard(explainer, title="First Dashboard", name="db1")
db1.to_yaml("dashboards/dashboard1/dashboard.yaml", dump_explainer=True, explainerfile_absolute_path=False)

db2 = ExplainerDashboard(explainer, title="Second dashboard", name="db2")
db2.to_yaml("dashboards/dashboard2/dashboard.yaml", dump_explainer=True, explainerfile_absolute_path=False)

hub = ExplainerHub([db1], 
                   add_dashboard_route=True, 
                   add_dashboard_pattern="dashboards/{}/dashboard.yaml")
    
hub.run()
```

Then, if I go to `http://localhost:8050/` I can see the first dashboard, cool. Then I go to `http://localhost:8050/add_dashboard/dashboard2`  I get this error:

```
ERROR: Failed to add dashboard!
127.0.0.1 - - [10/Oct/2024 15:45:36] "GET /add_dashboard/dashboard2 HTTP/1.1" 302 -
127.0.0.1 - - [10/Oct/2024 15:45:36] "GET / HTTP/1.1" 200 -
127.0.0.1 - - [10/Oct/2024 15:45:36] "GET /f"/static/jquery-3.5.1.slim.min.js" HTTP/1.1" 404 -
127.0.0.1 - - [10/Oct/2024 15:45:36] "GET /f"/static/bootstrap.min.js" HTTP/1.1" 404 -
127.0.0.1 - - [10/Oct/2024 15:45:36] "GET /static/bootstrap.min.css HTTP/1.1" 304 -
127.0.0.1 - - [10/Oct/2024 15:45:36] "GET /index/ HTTP/1.1" 200 -
127.0.0.1 - - [10/Oct/2024 15:45:36] "GET /index/assets/bootstrap.min.css?m=1728039482.0640564 HTTP/1.1" 304 -
127.0.0.1 - - [10/Oct/2024 15:45:36] "GET /index/_dash-layout HTTP/1.1" 200 -
127.0.0.1 - - [10/Oct/2024 15:45:36] "GET /index/_dash-dependencies HTTP/1.1" 200 -
```

I see that the error comes from [here](https://github.com/oegedijk/explainerdashboard/blob/master/explainerdashboard/dashboards.py#L2586). Not sure if I'm doing something wrong. Maybe I should use another parameter in the `to_yaml` method, but I don't see which one.

Thanks in advance for your insights!

**Comment by marcelo-mms** [comment]
I have an update. I changed the code in the explainerdashboard to print the exception:

```
Cannot find file: explainer.joblib
ERROR: Failed to add dashboard!
127.0.0.1 - - [10/Oct/2024 15:55:42] "GET /add_dashboard/dashboard2 HTTP/1.1" 302 -
127.0.0.1 - - [10/Oct/2024 15:55:42] "GET / HTTP/1.1" 200 -
127.0.0.1 - - [10/Oct/2024 15:55:42] "GET /f"/static/jquery-3.5.1.slim.min.js" HTTP/1.1" 404 -
127.0.0.1 - - [10/Oct/2024 15:55:42] "GET /f"/static/bootstrap.min.js" HTTP/1.1" 404 -
127.0.0.1 - - [10/Oct/2024 15:55:42] "GET /static/bootstrap.min.css HTTP/1.1" 304 -
127.0.0.1 - - [10/Oct/2024 15:55:42] "GET /index/ HTTP/1.1" 200 -
127.0.0.1 - - [10/Oct/2024 15:55:42] "GET /index/assets/bootstrap.min.css?m=1728039482.0640564 HTTP/1.1" 304 -
127.0.0.1 - - [10/Oct/2024 15:55:42] "GET /index/_dash-layout HTTP/1.1" 200 -
127.0.0.1 - - [10/Oct/2024 15:55:42] "GET /index/_dash-dependencies HTTP/1.1" 200 -
```

So ok, the path in the `yaml` file is wrong. I changed it to the relative path:

```yaml
dashboard:
  explainerfile: dashboards/dashboard2/explainer.joblib
  params:
    title: Second dashboard
    name: db2
    description: "This dashboard shows the workings of a fitted\n            machine\
      \ learning model, and explains its predictions."
    simple: false
    hide_header: false
    header_hide_title: false
    header_hide_selector: false
    header_hide_download: false
    hide_poweredby: false
    block_selector_callbacks: false
    pos_label: null
    fluid: true
    mode: dash
    width: 1000
    height: 800
    bootstrap: null
    external_stylesheets: null
    url_base_pathname: null
    routes_pathname_prefix: null
    requests_pathname_prefix: null
    responsive: true
    logins: null
    port: 8050
    importances: true
    model_summary: true
    contributions: true
    whatif: true
    shap_dependence: true
    shap_interaction: true
    decision_trees: true
    kwargs: {}
    tabs: null
```

And now I got a new error:

```
Building ExplainerDashboard..
The explainer object has no decision_trees property. so setting decision_trees=False...
Warning: calculating shap interaction values can be slow! Pass shap_interaction=False to remove interactions tab.
Generating layout...
Calculating dependencies...
Reminder: you can store the explainer (including calculated dependencies) with explainer.dump('explainer.joblib') and reload with e.g. ClassifierExplainer.from_file('explainer.joblib')
Registering callbacks...
Building ExplainerDashboard..
The setup method 'register_blueprint' can no longer be called on the application. It has already handled its first request, any changes will not be applied consistently.
Make sure all imports, decorators, functions, etc. needed to set up the application are done before running it.
ERROR: Failed to add dashboard!
127.0.0.1 - - [10/Oct/2024 15:56:22] "GET /add_dashboard/dashboard2 HTTP/1.1" 302 -
127.0.0.1 - - [10/Oct/2024 15:56:22] "GET / HTTP/1.1" 200 -
127.0.0.1 - - [10/Oct/2024 15:56:22] "GET /f"/static/jquery-3.5.1.slim.min.js" HTTP/1.1" 404 -
127.0.0.1 - - [10/Oct/2024 15:56:22] "GET /f"/static/bootstrap.min.js" HTTP/1.1" 404 -
127.0.0.1 - - [10/Oct/2024 15:56:22] "GET /static/bootstrap.min.css HTTP/1.1" 304 -
127.0.0.1 - - [10/Oct/2024 15:56:22] "GET /f"/static/bootstrap.min.js" HTTP/1.1" 404 -
127.0.0.1 - - [10/Oct/2024 15:56:22] "GET /index/ HTTP/1.1" 200 -
127.0.0.1 - - [10/Oct/2024 15:56:22] "GET /index/assets/bootstrap.min.css?m=1728039482.0640564 HTTP/1.1" 304 -
127.0.0.1 - - [10/Oct/2024 15:56:22] "GET /index/_dash-layout HTTP/1.1" 200 -
127.0.0.1 - - [10/Oct/2024 15:56:22] "GET /index/_dash-dependencies HTTP/1.1" 200 -
```

Maybe I should use an older version of `dash`, I don't know. This is not my field 😅

**Pull request merged**